### PR TITLE
*: add check-task cmd; refine output format

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve? 

- add `check-task` command
- refine output with task config check response

### What is changed and how it works?

- add `check-task` to call `CheckTask` in DM-master
- refine JSON string in response

**before**:
```bash
» check-task task.shard.yaml
{
    "result": false,
    "msg": "fail to check synchronization configuration with type CheckFailed:\n check was failed, please see detail\n detail: {\n\t\t\"results\": [\n\t\t\t{\n\t\t\t\t\"id\": 9,\n\t\t\t\t\"name\": \"table structure compatibility check\",\n\t\t\t\t\"desc\": \"check compatibility of table structure\",\n\t\t\t\t\"state\": \"success\",\n\t\t\t\t\"errorMsg\": \"\",\n\t\t\t\t\"instruction\": \"\",\n\t\t\t\t\"extra\": \"address of db instance - 127.0.0.1:3308\"\n\t\t\t},\n\t\t\t{\n\t\t\t\t\"id\": 1,\n\t\t\t\t\"name\": \"mysql_binlog_format\",\n\t\t\t\t\"desc\": \"check whether mysql binlog_format is ROW\",\n\t\t\t\t\"state\": \"success\",\n\t\t\t\t\"errorMsg\": \"\",\n\t\t\t\t\"instruction\": \"\",\n\t\t\t\t\"extra\": \"address of db instance - 127.0.0.1:3307\"\n\t\t\t},\n\t\t\t{\n\t\t\t\t\"id\": 3,\n\t\t\t\t\"name\": \"source_db_privilege\",\n\t\t\t\t\"desc\": \"check privileges of source DB\",\n\t\t\t\t\"state\": \"success\",\n\t\t\t\t\"errorMsg\": \"\",\n\t\t\t\t\"instruction\": \"\",\n\t\t\t\t\"extra\": \"address of db instance - 127.0.0.1:3307\"\n\t\t\t},\n\t\t\t{\n\t\t\t\t\"id\": 5,\n\t\t\t\t\"name\": \"mysql_binlog_enable\",\n\t\t\t\t\"desc\": \"check whether mysql binlog is enabled\",\n\t\t\t\t\"state\": \"success\",\n\t\t\t\t\"errorMsg\": \"\",\n\t\t\t\t\"instruction\": \"\",\n\t\t\t\t\"extra\": \"address of db instance - 127.0.0.1:3308\"\n\t\t\t},\n\t\t\t{\n\t\t\t\t\"id\": 8,\n\t\t\t\t\"name\": \"source_db_privilege\",\n\t\t\t\t\"desc\": \"check privileges of source DB\",\n\t\t\t\t\"state\": \"success\",\n\t\t\t\t\"errorMsg\": \"\",\n\t\t\t\t\"instruction\": \"\",\n\t\t\t\t\"extra\": \"address of db instance - 127.0.0.1:3308\"\n\t\t\t},\n\t\t\t{\n\t\t\t\t\"id\": 0,\n\t\t\t\t\"name\": \"mysql_binlog_enable\",\n\t\t\t\t\"desc\": \"check whether mysql binlog is enabled\",\n\t\t\t\t\"state\": \"success\",\n\t\t\t\t\"errorMsg\": \"\",\n\t\t\t\t\"instruction\": \"\",\n\t\t\t\t\"extra\": \"address of db instance - 127.0.0.1:3307\"\n\t\t\t},\n\t\t\t{\n\t\t\t\t\"id\": 2,\n\t\t\t\t\"name\": \"mysql_binlog_row_image\",\n\t\t\t\t\"desc\": \"check whether mysql binlog_row_image is FULL\",\n\t\t\t\t\"state\": \"success\",\n\t\t\t\t\"errorMsg\": \"\",\n\t\t\t\t\"instruction\": \"\",\n\t\t\t\t\"extra\": \"address of db instance - 127.0.0.1:3307\"\n\t\t\t},\n\t\t\t{\n\t\t\t\t\"id\": 4,\n\t\t\t\t\"name\": \"table structure compatibility check\",\n\t\t\t\t\"desc\": \"check compatibility of table structure\",\n\t\t\t\t\"state\": \"success\",\n\t\t\t\t\"errorMsg\": \"\",\n\t\t\t\t\"instruction\": \"\",\n\t\t\t\t\"extra\": \"address of db instance - 127.0.0.1:3307\"\n\t\t\t},\n\t\t\t{\n\t\t\t\t\"id\": 6,\n\t\t\t\t\"name\": \"mysql_binlog_format\",\n\t\t\t\t\"desc\": \"check whether mysql binlog_format is ROW\",\n\t\t\t\t\"state\": \"success\",\n\t\t\t\t\"errorMsg\": \"\",\n\t\t\t\t\"instruction\": \"\",\n\t\t\t\t\"extra\": \"address of db instance - 127.0.0.1:3308\"\n\t\t\t},\n\t\t\t{\n\t\t\t\t\"id\": 10,\n\t\t\t\t\"name\": \"sharding table consistency check\",\n\t\t\t\t\"desc\": \"check consistency of sharding table structures\",\n\t\t\t\t\"state\": \"fail\",\n\t\t\t\t\"errorMsg\": \"column length mismatch (2 vs 1)\\n table `shard_db_2`.`shard_table_2`\\ncolumns c1 int(11)\\nc2 int(11)\\n\\ntable`shard_db_1`.`shard_table_1`\\ncolumns c1 int(11)\\n\",\n\t\t\t\t\"instruction\": \"please set same table structure for sharding tables\",\n\t\t\t\t\"extra\": \"instance 127.0.0.1:3307 on sharding `shard_db`.`shard_table`\"\n\t\t\t},\n\t\t\t{\n\t\t\t\t\"id\": 7,\n\t\t\t\t\"name\": \"mysql_binlog_row_image\",\n\t\t\t\t\"desc\": \"check whether mysql binlog_row_image is FULL\",\n\t\t\t\t\"state\": \"success\",\n\t\t\t\t\"errorMsg\": \"\",\n\t\t\t\t\"instruction\": \"\",\n\t\t\t\t\"extra\": \"address of db instance - 127.0.0.1:3308\"\n\t\t\t}\n\t\t],\n\t\t\"summary\": {\n\t\t\t\"passed\": false,\n\t\t\t\"total\": 11,\n\t\t\t\"successful\": 10,\n\t\t\t\"failed\": 1,\n\t\t\t\"warning\": 0\n\t\t}\n\t}"
}
```

**after**:
```
» check-task task.shard.yaml
{
    "result": false,
    "msg": "fail to check synchronization configuration with type CheckFailed: check was failed, please see detail
    	detail: {
		"results": [
			{
				"id": 8,
				"name": "source_db_privilege",
				"desc": "check privileges of source DB",
				"state": "success",
				"errorMsg": "",
				"instruction": "",
				"extra": "address of db instance - 127.0.0.1:3308"
			},
			{
				"id": 3,
				"name": "source_db_privilege",
				"desc": "check privileges of source DB",
				"state": "success",
				"errorMsg": "",
				"instruction": "",
				"extra": "address of db instance - 127.0.0.1:3307"
			},
			{
				"id": 9,
				"name": "table structure compatibility check",
				"desc": "check compatibility of table structure",
				"state": "success",
				"errorMsg": "",
				"instruction": "",
				"extra": "address of db instance - 127.0.0.1:3308"
			},
			{
				"id": 6,
				"name": "mysql_binlog_format",
				"desc": "check whether mysql binlog_format is ROW",
				"state": "success",
				"errorMsg": "",
				"instruction": "",
				"extra": "address of db instance - 127.0.0.1:3308"
			},
			{
				"id": 0,
				"name": "mysql_binlog_enable",
				"desc": "check whether mysql binlog is enabled",
				"state": "success",
				"errorMsg": "",
				"instruction": "",
				"extra": "address of db instance - 127.0.0.1:3307"
			},
			{
				"id": 1,
				"name": "mysql_binlog_format",
				"desc": "check whether mysql binlog_format is ROW",
				"state": "success",
				"errorMsg": "",
				"instruction": "",
				"extra": "address of db instance - 127.0.0.1:3307"
			},
			{
				"id": 10,
				"name": "sharding table consistency check",
				"desc": "check consistency of sharding table structures",
				"state": "fail",
				"errorMsg": "column length mismatch (1 vs 2)\n table `shard_db_1`.`shard_table_1`\ncolumns c1 int(11)\n\ntable`shard_db_2`.`shard_table_2`\ncolumns c1 int(11)\nc2 int(11)\n",
				"instruction": "please set same table structure for sharding tables",
				"extra": "instance 127.0.0.1:3308 on sharding `shard_db`.`shard_table`"
			},
			{
				"id": 4,
				"name": "table structure compatibility check",
				"desc": "check compatibility of table structure",
				"state": "success",
				"errorMsg": "",
				"instruction": "",
				"extra": "address of db instance - 127.0.0.1:3307"
			},
			{
				"id": 5,
				"name": "mysql_binlog_enable",
				"desc": "check whether mysql binlog is enabled",
				"state": "success",
				"errorMsg": "",
				"instruction": "",
				"extra": "address of db instance - 127.0.0.1:3308"
			},
			{
				"id": 2,
				"name": "mysql_binlog_row_image",
				"desc": "check whether mysql binlog_row_image is FULL",
				"state": "success",
				"errorMsg": "",
				"instruction": "",
				"extra": "address of db instance - 127.0.0.1:3307"
			},
			{
				"id": 7,
				"name": "mysql_binlog_row_image",
				"desc": "check whether mysql binlog_row_image is FULL",
				"state": "success",
				"errorMsg": "",
				"instruction": "",
				"extra": "address of db instance - 127.0.0.1:3308"
			}
		],
		"summary": {
			"passed": false,
			"total": 11,
			"successful": 10,
			"failed": 1,
			"warning": 0
		}
	}"
}
```

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)
 	- run `check-task`, `start-task` and `update-task` with unsupported upstream MySQL instances.

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@GregoryIan @amyangfei PTAL

**Comment by GregoryIan** [comment]
Good Job!

File: dm/master/server.go [hunk 1]
```diff
@@ -369,7 +369,7 @@ func (s *Server) UpdateTask(ctx context.Context, req *pb.UpdateTaskRequest) (*pb
 	if err != nil {
 		return &pb.UpdateTaskResponse{
 			Result: false,
-			Msg:    errors.ErrorStack(err),
```
Comment (GregoryIan) [inline dm/master/server.go]: why call `Cause()` here?

Comment (csuzhangxc) [reply dm/master/server.go]: not necessary. only to drop the annotated message to make check result clearer.

File: dm/ctl/common/util.go [hunk 2]
```diff
@@ -66,13 +66,83 @@ func PrintLines(format string, a ...interface{}) {
...
+			sr.Msg = placeholder // replace Msg with placeholder
+			marshaledStr, err = marshResponseToString(sr)
+			if err == nil {
+				replacedStr = strings.Replace(marshaledStr, placeholder, rawMsg, 1)
+			}
+		}
+	case *pb.UpdateTaskResponse:
+		sr := resp.(*pb.UpdateTaskResponse)
+		if strings.Contains(sr.Msg, subStr) {
+			found = true
```
Comment (GregoryIan) [inline dm/ctl/common/util.go]: extract these logic code to a function

File: dm/master/server.go (hunk 1 above)
Comment (GregoryIan) [reply dm/master/server.go]: ok

File: dm/ctl/common/util.go (hunk 2 above)
Comment (csuzhangxc) [reply dm/ctl/common/util.go]: a little hard to do that. these three proto.Message are `struct` but not `interface`, and we need to get/set their `Msg` field, so we need the exact type of them.

File: dm/ctl/master/start_task.go [hunk 3]
```diff
@@ -65,5 +67,7 @@ func startTaskFunc(cmd *cobra.Command, _ []string) {
 		return
 	}
 
-	common.PrettyPrintResponse(resp)
+	if !common.PrettyPrintResponseWithCheckTask(resp, checker.ErrorMsgHeader) {
```
Comment (GregoryIan) [inline dm/ctl/master/start_task.go]: no else？

File: dm/ctl/common/util.go (hunk 2 above)
Comment (GregoryIan) [reply dm/ctl/common/util.go]: is it ok?
```
switch sr := resp.(type) {
case xxx,yyy,zzz:
}
```

Comment (csuzhangxc) [reply dm/ctl/common/util.go]: not work. this can only **test** `type`, but we need **convert** type.

Comment (csuzhangxc) [reply dm/ctl/common/util.go]: updated to do the **convertion** in `switch` line, but cannot use multi cases in one Line (`case xxx,yyy,zzz:`).

**Comment by GregoryIan** [comment]
LGTM

File: dm/ctl/common/util.go [hunk 4]
```diff
@@ -66,13 +66,80 @@ func PrintLines(format string, a ...interface{}) {
...
+func PrettyPrintResponseWithCheckTask(resp proto.Message, subStr string) bool {
+	var (
+		err          error
+		found        bool
+		replacedStr  string
+		marshaledStr string
+		placeholder  = "PLACEHOLDER"
+	)
+	switch chr := resp.(type) {
+	case *pb.StartTaskResponse:
```
Comment (amyangfei) [inline dm/ctl/common/util.go]: we can use a unified `case *pb.StartTaskResponse, *pb.UpdateTaskResponse, *pb.CheckTaskResponse` as they have same logic code?

**Comment by amyangfei** [comment]
LGTM

**Pull request merged**